File: lib/gsd/tasks.rb
module GSD
  class  Tasks < Base
    get "/" do
      @tasks = Task.all

      erb :index
    end

    post "/tasks" do
      task = build_task

      if request.xhr?
        if task.save
          erb :task, {layout: false}, task: task
        else
          422
        end
      else
        if task.save
          session[:notification] = "Task was added successfully!"
        else
          session[:error] = "Missing data. Task could not be added."
        end
        redirect "/"
      end
    end

    patch "/tasks/:id/done" do |id|
      task = Task.find(id)
      task.done!


      if request.xhr?
        200
      else
        session[:notification] = "Task is now done! Congrats!"
        redirect "/"
      end
    end

    delete "/tasks/:id" do |id|
      task = Task.find(id)

      if request.xhr?
        if task.done?
          task.delete
          200
        else
          422
        end
      else
        if task.done?
          task.delete
          session[:notification] = "Task was deleted"
        else
          session[:error] = "Unfinished tasks cannot be deleted"
        end

        redirect "/"
      end
    end

    private

    def build_task
      Task.new(
        title: params[:task][:title],
        description: params[:task][:description],
        deadline: params[:task][:deadline]
      )
    end
  end
end

File: src/tasks.js
const express = require('express');
const router = express.Router();
const Task = require("./task");

router.get('/', function(req, res, next) {
  let notification = req.session.notification;
  if (notification) delete req.session.notification;
  let error = req.session.error;
  if (error) delete req.session.error;

  Task.all( (tasks) => res.render('index', { tasks: tasks, notification: notification, error: error }));
});

router.post('/', function(req, res, next) {
  const task = new Task(req.body.task);

  if (req.xhr) {
    if (task.save) {
      res.render('task', {layout: false, task: task});
    } else {
      res.status(422).end();
    }
  } else {
    if (task.save()) {
      req.session.notification = "Task was added successfully!";
    } else {
      req.session.error = "Missing data. Task could not be added.";
    }
    res.redirect("/tasks");
  }
});

router.patch("/:id/done", function(req, res, next) {
  console.log("Looking for task#" + req.params.id);
  Task.find(req.params.id, (task) => {
    console.log(task)
    task.markDone();

    if (req.xhr) {
      res.status(200).end();
    } else {
      req.session.notification = "Task is now done! Congrats!"
      res.redirect("/tasks");
    }
  });
});

router.delete("/:id", function(req, res, next) {
  Task.find(req.params.id, (task) => {
    if (req.xhr) {
      if (task.done) {
        task.destroy();
        res.status(200).end();
      } else {
        res.status(422).end();
      }
    } else {
      if (task.done) {
        task.destroy();
        req.session.notification = "Task was deleted"
      } else {
        req.session.error = "Unfinished tasks cannot be deleted"
      }
      res.redirect("/tasks");
    }
  });
});

module.exports = router;
